apps/prairielearn/src/lib/editors.js

```diff
@@ -123,27 +127,60 @@ export class Editor {
             await namedLocks.doWithLock(lockName, { timeout: 5000 }, async () => {
               const startGitHash = await getOrUpdateCourseCommitHash(this.course);
 
-              if (config.fileEditorUseGit) {
-                await cleanAndResetRepository(this.course, gitEnv, job);
-              }
+              if (!config.fileEditorUseGit) {
+                // If we are not using git (e.g., if we are running locally), then we:
+                //
+                // - Write changes to disk
+                // - Sync changes from disk
+                //
+                // Either the job ends with a thrown error or with the return statement.
 
-              try {
                 job.info('Write changes to disk');
+                job.data.saveAttempted = true;
                 await this.write();
-              } catch (err) {
-                if (config.fileEditorUseGit) {
-                  await cleanAndResetRepository(this.course, gitEnv, job);
-                }
+                job.data.saveSucceeded = true;
 
-                throw err;
-              }
-
-              if (!config.fileEditorUseGit) {
+                job.info('Sync changes from disk');
+                job.data.syncAttempted = true;
                 await syncCourseFromDisk(this.course, startGitHash, job);
+                job.data.syncSucceeded = true;
+
                 return;
               }
 
+              // If we are using git (e.g., if we are running in production), then we:
+              //
+              // - Pull from remote (then clean and reset)
+              // - Write changes to disk
+              // - Push to remote (then clean and reset)
+              // - Sync changes from disk
+              //
+              // If anything goes wrong in the pull, we error and exit.
+              //
+              // If anything goes wrong in the write or push, we make sure to clean/reset
+              // (removing changes made by this edit) and sync (because changes were made
+              // by the pull) before we error and exit.
+
+              job.info('Update to latest remote origin address');
+              await job.exec('git', ['remote', 'set-url', 'origin', this.course.repository], {
+                cwd: this.course.path,
+                env: gitEnv,
+              });
+
+              job.info('Fetch from remote git repository');
+              await job.exec('git', ['fetch'], {
+                cwd: this.course.path,
+                env: gitEnv,
+              });
```

Comment:
Should this be part of the `cleanAndResetRepository` function?

Reply:
I agree with this.